server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java

```diff
@@ -0,0 +1,603 @@
+                return ScheduledExecutors.Signal.STOP;
+              } else {
+                return ScheduledExecutors.Signal.REPEAT;
+              }
+            }
+          }
+      );
+    }
+  }
+
+  @Override
+  public void stop()
+  {
+    synchronized (lock) {
+      if (stopped) {
+        return;
+      }
+
+      stopped = true;
+
+      for (SegmentHolder holder : segmentsToDrop.values()) {
+        holder.requestCompleted(false, null);
+      }
+
+      for (SegmentHolder holder : segmentsToLoad.values()) {
+        holder.requestCompleted(false, null);
+      }
+
+      segmentsToDrop.clear();
+      segmentsToLoad.clear();
+      queuedSize.set(0L);
+      failedAssignCount.set(0);
+    }
+  }
+
+  @Override
+  public void loadSegment(DataSegment segment, LoadPeonCallback callback)
+  {
+    synchronized (lock) {
+      SegmentHolder holder = segmentsToLoad.get(segment);
+
+      if (holder == null) {
+        log.info("Server[%s] to load segment[%s] queued.", serverId, segment.getIdentifier());
+        segmentsToLoad.put(segment, new SegmentHolder(segment, LOAD, callback));
+        processingExecutor.execute(this::doSegmentManagement);
+      } else {
+        holder.addCallback(callback);
+      }
+    }
+  }
+
+  @Override
+  public void dropSegment(DataSegment segment, LoadPeonCallback callback)
+  {
+    synchronized (lock) {
+      SegmentHolder holder = segmentsToDrop.get(segment);
+
+      if (holder == null) {
+        log.info("Server[%s] to drop segment[%s] queued.", serverId, segment.getIdentifier());
+        segmentsToDrop.put(segment, new SegmentHolder(segment, DROP, callback));
+        processingExecutor.execute(this::doSegmentManagement);
+      } else {
+        holder.addCallback(callback);
+      }
+    }
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsToLoad()
+  {
+    return segmentsToLoad.keySet();
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsToDrop()
+  {
+    return segmentsToDrop.keySet();
+  }
+
+  @Override
+  public long getLoadQueueSize()
+  {
+    return queuedSize.get();
+  }
+
+  @Override
+  public int getAndResetFailedAssignCount()
+  {
+    return failedAssignCount.getAndSet(0);
+  }
+
+  @Override
+  public void markSegmentToDrop(DataSegment dataSegment)
+  {
+    segmentsMarkedToDrop.add(dataSegment);
+  }
+
+  @Override
+  public void unmarkSegmentToDrop(DataSegment dataSegment)
+  {
+    segmentsMarkedToDrop.remove(dataSegment);
+  }
+
+  @Override
+  public int getNumberOfSegmentsInQueue()
+  {
+    return segmentsToLoad.size();
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsMarkedToDrop()
+  {
+    return segmentsMarkedToDrop;
+  }
+
+  private class SegmentHolder
+  {
+    private final DataSegment segment;
+    private final DataSegmentChangeRequest changeRequest;
+    private final int type;
+    private final List<LoadPeonCallback> callbacks = Lists.newArrayList();
+
+    private SegmentHolder(
+        DataSegment segment,
+        int type,
```

Comment:
What's the point of this type thing?  We should either leverage polymorphism or completely eliminate it.  Using it sometimes and not other times gets confusing

Reply:
changed to use Polymorphism.